### PR TITLE
Implements docker firewall validation

### PR DESCRIPTION
**Pull request by b10n1k**
I added a small validation for the expected configuration of the firewall.
Briefly, makes sure that Docker firewall zone exists with an active docker0
interface and in turn this is not added on trusted zone. Run a container and
verify the iptables.

Applies firewall test in docker and docker_image
the tests implemented based on https://docs.docker.com/network/iptables/

- Related ticket: https://progress.opensuse.org/issues/89872
- Verification run: 
[OSD](https://openqa.suse.de/tests/overview?distri=sle&version=15-SP3&build=b10n1k%2Fos-autoinst-distri-opensuse%2312448)
[TW](https://openqa.opensuse.org/tests/overview?distri=opensuse&build=b10n1k%2Fos-autoinst-distri-opensuse%2312448&version=Tumbleweed)
[Leap](https://openqa.opensuse.org/tests/overview?distri=opensuse&build=b10n1k%2Fos-autoinst-distri-opensuse%2312448&version=15.3)
[sle-12-SP5](https://openqa.suse.de/tests/5954410)

### CODE REVIEW

File: tests/containers/docker_image.pm [hunk 1]
```diff
@@ -23,19 +24,21 @@ use Mojo::Base qw(consoletest);
...
+    my ($self) = @_;
     my ($image_names, $stable_names) = get_suse_container_urls();
     my ($running_version, $sp, $host_distri) = get_os_release;
     my $runtime = "docker";
 
     install_docker_when_needed($host_distri);
     allow_selected_insecure_registries(runtime => $runtime);
     scc_apply_docker_image_credentials() if (get_var('SCC_DOCKER_IMAGE'));
-
+    check_docker_firewall()              if (check_runtime_version(get_docker_version(), ">=20.10") && !($host_distri ne 'centos' || $host_distri ne 'ubuntu'));
```
Comment (Vogtinator) [inline tests/containers/docker_image.pm:41]: `!($host_distri ne 'centos' || $host_distri ne 'ubuntu'))` is always `true`. Did you mean

`!($host_distri eq 'centos' || $host_distri eq 'ubuntu'))`? AFAICT checking for openSUSE/SLE specifically makes more sense

File: lib/containers/utils.pm [hunk 2]
```diff
@@ -41,20 +41,56 @@ sub test_seccomp {
...
+}
 
+sub basic_container_tests {
+    my %args    = @_;
+    my $runtime = $args{runtime};
+    die "You must define the runtime!" unless $runtime;
     my $alpine_image_version = '3.6';
-    my $alpine               = "${prefix}alpine:$alpine_image_version";
-    my $hello_world          = "${prefix}hello-world";
+    my $alpine               = registry_url('alpine', '3.6');
```
Comment (Vogtinator) [inline lib/containers/utils.pm:92]:
```suggestion
    my $alpine               = registry_url('alpine', $alpine_image_version);
```

File: tests/containers/docker_image.pm (hunk 1 above)
Comment (pdostal) [reply tests/containers/docker_image.pm:41]: It makes no sense to run this on `docker.pm` as well as here. It's enough in `docker.pm`.

Comment (pdostal) [reply tests/containers/docker_image.pm:41]: I also consider routines `check_runtime_version` and `get_docker_version` as over-engineering.
One `script_output` condition should be enough.

File: lib/containers/utils.pm [hunk 3]
```diff
@@ -41,20 +41,56 @@ sub test_seccomp {
...
+}
 
+sub container_ip {
+    my ($container, $runtime) = @_;
+    my $ip = script_output "$runtime inspect $container --format='{{.NetworkSettings.IPAddress}}'";
+    record_info "$ip";
+    return $ip;
+}
+
+sub registry_url {
```
Comment (pdostal) [inline lib/containers/utils.pm:75]: I like this idea but it is in my opinion too big and should be in separate pull request or at least commit.

File: lib/containers/utils.pm [hunk 4]
```diff
@@ -41,20 +41,56 @@ sub test_seccomp {
     }
 }
 
-sub basic_container_tests {
-    my %args    = @_;
-    my $runtime = $args{runtime};
-    die "You must define the runtime!" unless $runtime;
+sub check_docker_firewall {
+    my $container_name = 'sut_container';
+    validate_script_output('ip a s docker0', sub { /state DOWN/ });
```
Comment (pdostal) [inline lib/containers/utils.pm:46]: Why do you check that the `docker0` is `down`? There might be other containers running.
If you are sure that when no containers are running the `docker0` is `down` then please test that.

File: lib/containers/utils.pm [hunk 5]
```diff
@@ -41,20 +41,56 @@ sub test_seccomp {
...
+    assert_script_run "firewall-cmd --list-all --zone=docker";
```
Comment (pdostal) [inline lib/containers/utils.pm:47]: We don't use `firewalld` on *<SLE15* but `SUSEfirewall2`.

File: lib/containers/utils.pm [hunk 6]
```diff
@@ -41,20 +41,56 @@ sub test_seccomp {
...
+    validate_script_output "firewall-cmd --list-interfaces --zone=docker",  sub { /docker0/ };
+    validate_script_output "firewall-cmd --list-interfaces --zone=trusted", sub { /^\s*$/ };
+    validate_script_output "iptables -L DOCKER-USER -nvx --line-numbers",   sub { /1.+all.+0\.0\.0\.0\/0\s+0\.0\.0\.0\/0/ };
```
Comment (pdostal) [inline lib/containers/utils.pm:50]: Please comment some example output.

File: lib/containers/utils.pm [hunk 7]
```diff
@@ -41,20 +41,56 @@ sub test_seccomp {
...
+    assert_script_run "docker run -id --rm --name $container_name -p 1234:1234 " . registry_url('alpine');
+    my $container_ip = container_ip($container_name, "docker");
+    validate_script_output "iptables -L DOCKER -nvx --line-numbers", sub { /1.+ACCEPT.+!docker0 docker0.+$container_ip\s+tcp dpt:1234/ };
```
Comment (pdostal) [inline lib/containers/utils.pm:53]: Please comment some example output.

**Review by pdostal — CHANGES_REQUESTED** [review]

File: tests/containers/docker_image.pm (hunk 1 above)
Comment (b10n1k) [reply tests/containers/docker_image.pm:41]: > It makes no sense to run this on `docker.pm` as well as here. It's enough in `docker.pm`.

i am not sure what do you mean here. i added initially to the docker.pm. But we have scheduled tests that run only the `containers_sle_image_on_sle_host` with each container build and do not have the docker.pm.

Comment (b10n1k) [reply tests/containers/docker_image.pm:41]: > I also consider routines `check_runtime_version` and `get_docker_version` as over-engineering.
> One `script_output` condition should be enough.

Maybe a bit. but i believe they make sense. Especially the `check_runtime_version`.

File: lib/containers/utils.pm (hunk 4 above)
Comment (b10n1k) [reply lib/containers/utils.pm:46]: `check_docker_firewall` is executed after `install_docker_when_needed`. in this case the most possible is that no containers are running. But i guess it adds an extra net if we provide a check. wdyt?

File: lib/containers/utils.pm (hunk 3 above)
Comment (b10n1k) [reply lib/containers/utils.pm:79]: moved [here b7f7e57](b7f7e57ac694b03ea7187fbc9738c6b9dc669e2e)

File: lib/containers/utils.pm (hunk 4 above)
Comment (b10n1k) [reply lib/containers/utils.pm]: i added a line for checking

File: lib/containers/utils.pm (hunk 5 above)
Comment (b10n1k) [reply lib/containers/utils.pm:48]: nice catch. check the update

File: tests/containers/docker_image.pm (hunk 1 above)
Comment (b10n1k) [reply tests/containers/docker_image.pm]: > `!($host_distri ne 'centos' || $host_distri ne 'ubuntu'))` is always `true`. Did you mean
> 
> `!($host_distri eq 'centos' || $host_distri eq 'ubuntu'))`? AFAICT checking for openSUSE/SLE specifically makes more sense

fixed

File: tests/containers/docker_image.pm [hunk 8]
```diff
@@ -23,19 +24,21 @@ use Mojo::Base qw(consoletest);
...
+    my ($self) = @_;
     my ($image_names, $stable_names) = get_suse_container_urls();
     my ($running_version, $sp, $host_distri) = get_os_release;
     my $runtime = "docker";
 
     install_docker_when_needed($host_distri);
     allow_selected_insecure_registries(runtime => $runtime);
     scc_apply_docker_image_credentials() if (get_var('SCC_DOCKER_IMAGE'));
-
+    check_docker_firewall() if (check_runtime_version(get_docker_version(), ">=20.10") && $self->firewall() eq 'firewalld' && (!($host_distri eq 'centos' || $host_distri eq 'ubuntu')));
```
Comment (Vogtinator) [inline tests/containers/docker_image.pm:41]: Can be simplified:

```suggestion
    check_docker_firewall() if (check_runtime_version(get_docker_version(), ">=20.10") && $self->firewall() eq 'firewalld' && $host_distri ne 'centos' && $host_distri ne 'ubuntu');
```

Comment (b10n1k) [reply tests/containers/docker_image.pm]: done

File: lib/containers/utils.pm [hunk 9]
```diff
@@ -41,20 +41,61 @@ sub test_seccomp {
...
 }
 
-sub basic_container_tests {
-    my %args    = @_;
-    my $runtime = $args{runtime};
-    die "You must define the runtime!" unless $runtime;
+sub check_docker_firewall {
+    my $container_name = 'sut_container';
+    my $running        = script_output qq(docker ps -a -q);
+    validate_script_output('ip a s docker0', sub { /state DOWN/ }) if $running eq '';
```
Comment (pdostal) [inline lib/containers/utils.pm:47]:
```suggestion
    my $running        = script_output qq(docker ps -a -q | wc -l);
    validate_script_output('ip a s docker0', sub { /state DOWN/ }) if $running ne 0;
```

File: lib/containers/utils.pm [hunk 10]
```diff
@@ -41,20 +41,61 @@ sub test_seccomp {
...
+    my ($container_name, $version_tag) = @_;
     my $registry = get_var('REGISTRY', 'docker.io');
     # Images from docker.io registry are listed without the 'docker.io/library/'
     # Images from custom registry are listed with the 'server/library/'
     # We also filter images the same way they are listed.
-    my $prefix = ($registry =~ /docker\.io/) ? "" : "$registry/library/";
+    my $repo = ($registry =~ /docker\.io/) ? "" : "$registry/library/";
+    return $repo                   unless $container_name;
+    return $repo . $container_name unless $version_tag;
+    return "$repo$container_name:$version_tag";
```
Comment (pdostal) [inline lib/containers/utils.pm:89]: Don't you wanna validate the repository URL?
It can end with or without slash.

File: tests/containers/docker_image.pm [hunk 11]
```diff
@@ -23,19 +24,21 @@ use Mojo::Base qw(consoletest);
...
+    my ($self) = @_;
     my ($image_names, $stable_names) = get_suse_container_urls();
     my ($running_version, $sp, $host_distri) = get_os_release;
     my $runtime = "docker";
 
     install_docker_when_needed($host_distri);
     allow_selected_insecure_registries(runtime => $runtime);
     scc_apply_docker_image_credentials() if (get_var('SCC_DOCKER_IMAGE'));
-
+    check_docker_firewall() if (check_runtime_version(get_docker_version(), ">=20.10") && $self->firewall() eq 'firewalld' && $host_distri ne 'centos' && $host_distri ne 'ubuntu');
```
Comment (pdostal) [inline tests/containers/docker_image.pm:41]: Why is this in both `docker.pm` as well as herer in `docker_image.pm`?

**Review by pdostal — COMMENTED** [review]
This is nice improvement! I have just few more questions.

File: tests/containers/docker_image.pm (hunk 11 above)
Comment (b10n1k) [reply tests/containers/docker_image.pm:41]: i answered on this [here](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/12448#discussion_r625141156)

File: lib/containers/utils.pm (hunk 10 above)
Comment (b10n1k) [reply lib/containers/utils.pm:89]: we can add a validation. but i dont see necessity for it. nothing has changed from the previous approach. and i am not sure what part you mean will end with/without "/". Can you give me an example where the problem will be?

File: lib/containers/utils.pm (hunk 9 above)
Comment (b10n1k) [reply lib/containers/utils.pm]: done

File: lib/containers/utils.pm (hunk 10 above)
Comment (pdostal) [reply lib/containers/utils.pm:89]: In [this job](https://openqa.suse.de/tests/5943067#settings) we have `REGISTRY=18.156.2.117:5000` which means that for the `hello-world` image this function will return `18.156.2.117:5000library/hello-world` and that is wrong.

Comment (pdostal) [reply lib/containers/utils.pm:89]: I would like to see the returns like this:
```perl
return $repo                                   unless $container_name;
return "$repo/$container_name" unless $version_tag;
return "$repo$container_name:$version_tag";
```

File: tests/containers/docker_image.pm (hunk 1 above)
Comment (pdostal) [reply tests/containers/docker_image.pm]: > > It makes no sense to run this on `docker.pm` as well as here. It's enough in `docker.pm`.
> 
> i am not sure what do you mean here. i added initially to the docker.pm. But we have scheduled tests that run only the `containers_sle_image_on_sle_host` with each container build and do not have the docker.pm.

True, for the Latest we have those. But we should then also have test sute for testing the runtime (e.g. `docker.pm`, `podman.pm`, `buildah.pm`). @jlausuch wdyt?

File: lib/containers/utils.pm (hunk 10 above)
Comment (b10n1k) [reply lib/containers/utils.pm:89]: check this `my $repo = ($registry =~ /docker\.io/) ? "" : "$registry/library/";` the registry is substituted and returns `18.156.2.117:5000library/`. Then for `hello-world` L88 should be used and this will return `18.156.2.117:5000library/hello-world`. i am pretty sure there is not a problem here.

> In [this job](https://openqa.suse.de/tests/5943067#settings) we have `REGISTRY=18.156.2.117:5000` which means that for the `hello-world` image this function will return `18.156.2.117:5000library/hello-world` and that is wrong.

Comment (pdostal) [reply lib/containers/utils.pm:89]:
```diff
- 18.156.2.117:5000library/hello-world
+ 18.156.2.117:5000library/hello-world
```

Comment (pdostal) [reply lib/containers/utils.pm:89]:
```diff
- 18.156.2.117:5000library/hello-world
+ 18.156.2.117:5000/library/hello-world
```

Comment (b10n1k) [reply lib/containers/utils.pm:89]: where from?

Comment (pdostal) [reply lib/containers/utils.pm:89]: image

File: lib/containers/utils.pm [hunk 12]
```diff
@@ -41,20 +41,61 @@ sub test_seccomp {
...
+sub container_ip {
+    my ($container, $runtime) = @_;
+    my $ip = script_output "$runtime inspect $container --format='{{.NetworkSettings.IPAddress}}'";
+    record_info "$ip";
+    return $ip;
+}
+
+sub registry_url {
+    my ($container_name, $version_tag) = @_;
     my $registry = get_var('REGISTRY', 'docker.io');
```
Comment (pdostal) [inline lib/containers/utils.pm:82]:
```suggestion
    my $registry = get_var('REGISTRY', 'docker.io');
    $registry =~ s{/$}{}; # remove trailing slash at the end of string
```

File: lib/containers/utils.pm [hunk 13]
```diff
@@ -41,20 +41,61 @@ sub test_seccomp {
...
     # Images from docker.io registry are listed without the 'docker.io/library/'
     # Images from custom registry are listed with the 'server/library/'
     # We also filter images the same way they are listed.
-    my $prefix = ($registry =~ /docker\.io/) ? "" : "$registry/library/";
+    my $repo = ($registry =~ /docker\.io/) ? "" : "$registry/library";
+    return $repo unless $container_name;
```
Comment (pdostal) [inline lib/containers/utils.pm:87]:
```suggestion
    return $registry unless $container_name;
```
When you don't specify image name you care just about the registry URL.

File: lib/containers/utils.pm (hunk 12 above)
Comment (b10n1k) [reply lib/containers/utils.pm:82]: added

File: lib/containers/utils.pm (hunk 13 above)
Comment (b10n1k) [reply lib/containers/utils.pm:89]: i am not sure about this. this is what i had thought as well in first place but the fact is that this would be useful only in one place in `lib/containers/common.pm` which for simplicity can stay as it is. Now i am thinking if even we ever need this return if it is not used. late me take another look.

Comment (pdostal) [reply lib/containers/utils.pm:89]: As you now created this routine it should be used wherever possible and it should be as universal as possible.

File: lib/containers/utils.pm (hunk 10 above)
Comment (b10n1k) [reply lib/containers/utils.pm]: solved

File: lib/containers/utils.pm (hunk 13 above)
Comment (b10n1k) [reply lib/containers/utils.pm]: done

**Comment by b10n1k** [comment]
messed up goood time

**Comment by pdostal** [comment]
Will you reopen / recreate @b10n1k?
